ada_feeding_perception/ada_feeding_perception/face_detection.py

```diff
@@ -0,0 +1,388 @@
+                    ),
+                ),
+                (
+                    "landmark_model_name",
+                    None,
+                    ParameterDescriptor(
+                        name="landmark_model_name",
+                        type=ParameterType.PARAMETER_STRING,
+                        description="The name of the face detection model checkpoint to use",
+                        read_only=True,
+                    ),
+                ),
+                (
+                    "landmark_model_base_url",
+                    None,
+                    ParameterDescriptor(
+                        name="landmark_model_base_url",
+                        type=ParameterType.PARAMETER_STRING,
+                        description=(
+                            "The URL to download the model checkpoint from "
+                            "if it is not already downloaded"
+                        ),
+                        read_only=True,
+                    ),
+                ),
+                (
+                    "model_dir",
+                    None,
+                    ParameterDescriptor(
+                        name="model_dir",
+                        type=ParameterType.PARAMETER_STRING,
+                        description=(
+                            "The location of the directory where the model "
+                            "checkpoint is / should be stored"
+                        ),
+                        read_only=True,
+                    ),
+                ),
+            ],
+        )
+
+    def toggle_face_detection_callback(
+        self, request: SetBool.Request, response: SetBool.Response
+    ) -> SetBool.Response:
+        """
+        Callback function for the toggle_face_detection service. Safely toggles
+        the face detection on or off depending on the request.
+        """
+        self.get_logger().info(f"Incoming service request. data: {request.data}")
+        with self.is_on_lock:
+            self.is_on = request.data
+        response.success = request.data
+        return response
+
+    def camera_info_callback(self, msg: CameraInfo):
+        """
+        Callback function for the toggle_face_detection service. Safely toggles
+        the face detection on or off depending on the request.
+        """
+        self.camera_matrix = msg.k
+
+    def depth_callback(self, msg: CompressedImage):
+        """
+        Callback function for depth images. If face_detection is on, this
+        function publishes the 3d location of the mouth on the /face_detection
+        topic in the camera_depth_optical_frame
+        """
+
+        # Collect last ten depth images (depth_buffer is a collections.deque of max length 10)
+        self.depth_buffer.append(msg)
+        with self.is_on_lock:
+            is_on = self.is_on
+
+        # Check if face detection is on
+        if is_on:
+            face_detection_msg = FaceDetection()
+            if self.is_face_detected:
```

Comment:
I think `self.is_face_detected` should also be protected by a lock, since it can be accessed across multiple threads. Same with the other attributes that are set in `color_callback` and used here: `self.img_mouth_center`, `self.img_timestamp`, and `self.img_mouth_points`. The main reason for this is to ensure that all these attributes are read from a message of the same timestamp. As it stands, if a call to `color_callback` is being run at the same time as this `depth_callback`, you might get the `img_mouth_center` from the previous call but the `img_mouth_points` from the new call, which is undesirable.

Reply:
To address this, I'd suggest having one lock `self.mouth_detection_lock` to handle all instance attributes that are set in `color_callback`. In `color_callback`, do all your computations with local variables, and only at the end of the function you should acquire the lock and set all the instance attributes. Similarly, at the top of `depth_callback` you should acquire the lock, store all instance attributes as local variables, and then only use the local variables for the rest of the function.